core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala

```diff
@@ -473,6 +473,146 @@ class DAGSchedulerSuite
     assertDataStructuresEmpty()
   }
 
+  test("Multiple consecutive stage failures should lead to stage being aborted.") {
+    // Create a new Listener to confirm that the listenerBus sees the JobEnd message
+    // when we abort the stage. This message will also be consumed by the EventLoggingListener
+    // so this will propagate up to the user.
+    var ended = false
+    var jobResult : JobResult = null
+    class EndListener extends SparkListener {
+      override def onJobEnd(jobEnd: SparkListenerJobEnd): Unit = {
+        jobResult = jobEnd.jobResult
+        ended = true
+      }
+    }
+
+    sc.listenerBus.addListener(new EndListener())
+
+    val shuffleMapRdd = new MyRDD(sc, 2, Nil)
+    val shuffleDep = new ShuffleDependency(shuffleMapRdd, null)
+    val shuffleId = shuffleDep.shuffleId
+    val reduceRdd = new MyRDD(sc, 2, List(shuffleDep))
+    submit(reduceRdd, Array(0, 1))
+
+    complete(taskSets(0), Seq(
+      (Success, makeMapStatus("hostA", 2)),
+      (Success, makeMapStatus("hostB", 2))))
+
+    for (x <- 1 to Stage.MAX_STAGE_FAILURES) {
+      // the 2nd ResultTask failed
+      complete(taskSets(1), Seq(
+        (Success, 42),
+        (FetchFailed(makeBlockManagerId("hostA"), shuffleId, 0, 0, "ignored"), null)))
```

Comment:
It might also be helpful to realize that you get this behavior with `taskSets` is because within this test, there is a custom `TaskScheduler`, that just keeps appending to `taskSets` anytime a new one is added.  Nothing is ever removed, [see here](https://github.com/apache/spark/blob/master/core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala#L89).